### ISSUE TITLE
An issue on implementation of SemanticConnectivityloss

### ISSUE THREAD

**Issue by rose-jinyang** [PaddlePaddle/PaddleSeg#1798]
Hi @LutaoChu and @dreamer121121
In the forward function of the SemanticConnectivityLoss implementation(https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.4/paddleseg/models/losses/semantic_connectivity_loss.py), are the parameter logits the probabilities map or logit map without applying softmax?
BTW:
![image](https://user-images.githubusercontent.com/47862419/155523645-e41fe90b-91ff-4b39-8237-814aa4e76e54.png)
There is an issue with the above function implementation.
pred_conn has labels greater than pred_num_conn value.
That's because pred_num_conn is limited as smaller than 10.
![image](https://user-images.githubusercontent.com/47862419/155523705-2defcd2b-d1e2-437a-b1b6-12781eb34cdb.png)
This results in an issue in the below line.
pred_conn = F.one_hot(pred_conn.long(), pred_num_conn)

How can we fix this issue?

**Comment by LutaoChu** [comment]
Hi, Logits is logit map without applying softmax.

Regarding the above issue, it looks like there is indeed a problem. We need to slice pred_conn. 
But it is very strange that I did not find any errors during the actual training process. Can you provide your error message?

**Comment by LutaoChu** [comment]
I tested it for paddle. If the range of values in pred_conn exceeds the pred_num_conn value, F.one_hot can still be calculated

**Comment by rose-jinyang** [comment]
In fact, I am using this loss in PyTorch.
In PyTorch, F.one_hot does not handle such case.

**Comment by rose-jinyang** [comment]
In case of using Paddle, how will the values in pred_conn exceeds the pred_num_conn value be processed?

**Comment by LutaoChu** [comment]
> In fact, I am using this loss in PyTorch. In PyTorch, F.one_hot does not handle such case.

You can try it as the following
![image](https://user-images.githubusercontent.com/30695251/155719577-7fb0d963-2364-46f7-8a9e-e2a1c06ed396.png)

**Comment by LutaoChu** [comment]
> In case of using Paddle, how will the values in pred_conn exceeds the pred_num_conn value be processed?

Exceeded values will be ignored. The corresponding one hot codes returned are all 0
```
import paddle
import paddle.nn.functional as F

a = paddle.randint(0, 10, (4, 4))
b = F.one_hot(a, 5)
print(a)
print(b)
```
Output
```
Tensor(shape=[4, 4], dtype=int64, place=CUDAPlace(0), stop_gradient=True,
       [[7, 2, 3, 3],
        [8, 6, 2, 6],
        [8, 1, 7, 4],
        [4, 1, 8, 6]])
Tensor(shape=[4, 4, 5], dtype=float32, place=CUDAPlace(0), stop_gradient=True,
       [[[0., 0., 0., 0., 0.],
         [0., 0., 1., 0., 0.],
         [0., 0., 0., 1., 0.],
         [0., 0., 0., 1., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 1., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 1., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 1.]],

        [[0., 0., 0., 0., 1.],
         [0., 1., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]])
```

**Comment by rose-jinyang** [comment]
I've got the below issue with the above code.

![image](https://user-images.githubusercontent.com/47862419/155739266-5e57a965-da52-49c4-81aa-173e7ca40fb0.png)

**Comment by rose-jinyang** [comment]
What is your paddlepaddle version? Mine is paddlepaddle 2.2.2.

**Comment by LutaoChu** [comment]
Mine is 2.2.1
Please give your running command and I'll repeat the problem

**Comment by rose-jinyang** [comment]
I've installed paddle 2.2.1.
But the issue appears still.

![image](https://user-images.githubusercontent.com/47862419/156101743-d82634a4-e33a-4b2d-a685-645d7033202a.png)

**Comment by LutaoChu** [comment]
I have no problem here
What is your running command?

**Comment by rose-jinyang** [comment]
import paddle
import paddle.nn.functional as F

a = paddle.randint(0, 10, (4, 4))
b = F.one_hot(a, 5)
print(a)
print(b)

This is the script.

**Comment by LutaoChu** [comment]
So strange. I can run it
![image](https://user-images.githubusercontent.com/30695251/156334401-d53d9b99-5ef7-45da-bafd-a87b56c08283.png)

**Comment by rose-jinyang** [comment]
The difference is only that I tried it on CPU without GPU.

**Comment by LutaoChu** [comment]
I have the same problem with CPU. This is an OP bug in paddle. You can submit the issue here  https://github.com/PaddlePaddle/Paddle

**Comment by LutaoChu** [comment]
> > In fact, I am using this loss in PyTorch. In PyTorch, F.one_hot does not handle such case.
> 
> You can try it as the following ![image](https://user-images.githubusercontent.com/30695251/155719577-7fb0d963-2364-46f7-8a9e-e2a1c06ed396.png)

If you only have CPU, you can modify it according to this

**Comment by rose-jinyang** [comment]
One of Paddle developers recognized this issue on GPU.
I think that your loss implementation should be updated under consideration of this point.

![image](https://user-images.githubusercontent.com/47862419/157029594-531df39c-24cd-4337-b422-d84bc4a8e02d.png)

**Comment by rose-jinyang** [comment]
Could u provide the implementation of this loss in PyTorch.
That's because some issues appears when running this loss implemented by myself in PyTorch.

**Comment by LutaoChu** [comment]
In fact, I think that the input element can support greater than class_num is a feature, not a bug.

I will try the pytorch loss implementation

**Comment by rose-jinyang** [comment]
Hi @LutaoChu 
Did u implement PyTorch version of this loss?

**Comment by LutaoChu** [comment]
Hi，I fix this bug on CPU with Paddle. You can use it after the PR is merged
